### PR TITLE
Hyperparam tuning pipeline

### PR DESCRIPTION
**Pull request by jzmujdzin**

### CODE REVIEW

File: ticket_upgrade_prediction/hyperparam_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+from sklearn.preprocessing import StandardScaler
+
+from ticket_upgrade_prediction.pipeline import Pipeline as df_pipeline
+
+warnings.simplefilter(action="ignore", category=SettingWithCopyWarning)
+from sklearn.exceptions import DataConversionWarning
+
+warnings.filterwarnings(action="ignore", category=DataConversionWarning)
+warnings.simplefilter(action="ignore", category=FutureWarning)
+warnings.simplefilter("ignore", UserWarning)
```
Comment (kuba1302) [inline ticket_upgrade_prediction/hyperparam_pipeline.py]: Why do we ignore warnings instead of handling them?

File: ticket_upgrade_prediction/hyperparam_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+from typing import Tuple
+
+
+class HyperparamPipeline:
+    def __init__(
+        self,
+        X: pd.DataFrame,
+        y: pd.Series,
+        model: str,
```
Comment (kuba1302) [inline ticket_upgrade_prediction/hyperparam_pipeline.py]: I think documentation of the class should show possible values of the `model` variable, so that one would not have to look into `determine_model` method's code, to check which model is possible

File: ticket_upgrade_prediction/hyperparam_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+
+    def search_for_params(
+        self, searching_algo: str = "random", n_splits: int = 5, **kwargs
+    ) -> None:
+        logger.info(
+            f"starting to serach for params using {searching_algo} search"
+        )
+        if searching_algo == "random":
+            self.optimize_hypers_using_random_search(
+                n_splits, kwargs["n_iters"]
```
Comment (kuba1302) [inline ticket_upgrade_prediction/hyperparam_pipeline.py:76]: If we have to use chosen argument in `kwargs`, it should be at least documented. The way it is done now, one don't know what kwarg should be provided, without looking into function code

File: ticket_upgrade_prediction/hyperparam_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+        if self.metric == "accuracy_score":
+            return accuracy_score(y_pred, y_true)
+        elif self.metric == "balanced_accuracy":
+            return balanced_accuracy(y_pred, y_true)
+        elif self.metric == "precision":
+            return precision_score(y_pred, y_true)
+        elif self.metric == "recall":
+            return recall(y_pred, y_true)
+        elif self.metric == "roc_auc":
+            return roc_auc_score(y_pred, y_true)
```
Comment (kuba1302) [inline ticket_upgrade_prediction/hyperparam_pipeline.py]: `roc_auc` takes **probabilites** instead of labels. In this way, it will not work

File: ticket_upgrade_prediction/hyperparam_pipeline.py (hunk 2 above)
Comment (kuba1302) [inline ticket_upgrade_prediction/hyperparam_pipeline.py]: `model_type/name` would be more clear. We usually use `model` for a actual model class

File: ticket_upgrade_prediction/hyperparam_pipeline.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+        test_index: np.ndarray,
+        iteration_params: dict,
+    ) -> float:
+        model = self.determine_model(iteration_params)
+        X_train, X_test, y_train, y_test = self.get_scaled_train_and_test_sets(
+            train_index, test_index
+        )
+        model.fit(X_train, y_train)
+        # predictions = model.predict_proba(X_test)[:, 1] if self.classification else model.predict(X_test)
+        predictions = model.predict(X_test)
```
Comment (kuba1302) [inline ticket_upgrade_prediction/hyperparam_pipeline.py]: Why dont You use `Evaluator` class here?
I have updated `Metrics` dataclass here https://github.com/kuba1302/ticket-upgrade-prediction/pull/15, 
so it will be easier to choose metric from string.
You will be able to do something like this: 

```
evaluator = Evaluator(model=model, X=X_test, y=y_test)
metric = evaluator.get_all_metrics()
return metric.get_metric_from_string("roc_auc")
```

**Review by kuba1302 — CHANGES_REQUESTED** [review]

File: ticket_upgrade_prediction/hyperparam_pipeline.py (hunk 1 above)
Comment (jzmujdzin) [reply ticket_upgrade_prediction/hyperparam_pipeline.py]: upgrading packages fixes most of the errors. cant get rid of settingwithcopywarning, and i dont think its possible to get rid of it

File: ticket_upgrade_prediction/hyperparam_pipeline.py (hunk 2 above)
Comment (jzmujdzin) [reply ticket_upgrade_prediction/hyperparam_pipeline.py]: added doc to HyperparamPipeline class

File: ticket_upgrade_prediction/hyperparam_pipeline.py (hunk 3 above)
Comment (jzmujdzin) [reply ticket_upgrade_prediction/hyperparam_pipeline.py:76]: added to doc

File: ticket_upgrade_prediction/hyperparam_pipeline.py (hunk 2 above)
Comment (jzmujdzin) [reply ticket_upgrade_prediction/hyperparam_pipeline.py]: renamed model to model_type

File: ticket_upgrade_prediction/hyperparam_pipeline.py (hunk 4 above)
Comment (jzmujdzin) [reply ticket_upgrade_prediction/hyperparam_pipeline.py]: reworked script to use Evaluator class

File: ticket_upgrade_prediction/hyperparam_pipeline.py (hunk 5 above)
Comment (jzmujdzin) [reply ticket_upgrade_prediction/hyperparam_pipeline.py]: reworked script to use Evaluator class

**Review by kuba1302 — APPROVED** [review]
🚀
